ddsc/core/filedownloader.py

```diff
@@ -0,0 +1,175 @@
+"""
+Downloads a file based on ranges.
+"""
+import math
+import tempfile
+import requests
+from multiprocessing import Process, Queue
+from ddsc.core.util import ProgressQueue, wait_for_processes
+
+DOWNLOAD_FILE_CHUNK_SIZE = 20 * 1024 * 1024
+MIN_DOWNLOAD_CHUNK_SIZE = DOWNLOAD_FILE_CHUNK_SIZE
+
+
+class FileDownloader(object):
+    """
+    Downloads a file using a number of worker processes who download different ranges.
+    Creates an empty file.
+    Each worker seeks to their spot and streams the data from their url data into the file.
+    """
+    def __init__(self, config, remote_file, url_parts, path, watcher):
+        """
+        Setup details on what to download and watcher to notify of progress.
+        :param config: Config: configuration settings for download (number workers)
+        :param remote_file: RemoteFile: info about the file we are downloading
+        :param url_parts: dictionary of fields related to url ('http_verb','host','http_headers')
+        :param path: str: path to where we will save the file
+        :param watcher: ProgressPrinter: we notify of our progress
+        """
+        self.config = config
+        self.remote_file = remote_file
+        self.file_size = remote_file.size
+        self.url_parts = url_parts
+        self.path = path
+        self.watcher = watcher
+        self.file_parts = []
+
+    @property
+    def http_verb(self):
+        verb = self.url_parts['http_verb']
+        if verb != 'GET':
+            raise ValueError("Unsupported download method: {}".format(verb))
+        return verb
+
+    @property
+    def host(self):
```

Comment:
It looks like host and url properties are always used together. And `url` is actually the path. What about removing `host` and changing `url` to `return self.url_parts['host'] + self.url_parts['url']` and noting the convention